### PR TITLE
attach pst (Process Start Time) to all messages

### PR DESCRIPTION
**Pull request by rmg**
In combination with the process id and a supervisor identifier, this
can be used to create a unique identifier which simplifies a lot of
logic at higher abstractions.

Depends on strongloop/strong-cluster-control#49

/cc @kraman this should give us what we need from supervisor

### CODE REVIEW

File: lib/runctl.js [hunk 1]
```diff
@@ -150,9 +154,15 @@ function onRequest(req, callback) {
...
   } else if (cmd === 'fork') {
-    var child = cluster.fork();
-    rsp.workerID = child.workerID;
-    rsp.processID = child.process.pid;
+    rsp = null;
+    cluster.fork().once('fork', function(worker) {
+      var rsp = {
+        workerID: worker.id,
+        processID: worker.process.pid,
+        pst: worker.startTime,
```
Comment (sam-github) [inline lib/runctl.js]: Ignored by fork command... so not really necessary, but does no harm.

File: lib/status-wd.js [hunk 2]
```diff
@@ -21,10 +22,11 @@ module.exports = function(parentCtl) {
 
   if (!parentCtl) return;
 
-  cluster.on('fork', function(worker) {
+  master.on('fork', function(worker) {
     worker.on('message', function(msg) {
       if (msg.cmd !== 'status:wd') return;
       debug('master received: %j', msg);
+      msg.pst = worker.startTime;
```
Comment (sam-github) [inline lib/status-wd.js]: `.pst` property doesn't need to be part of the status:wd message, does it? Or are you trying to make sure `pid/wid/pst` triplet is every message that contains worker-identifying information?

File: lib/traces.js [hunk 3]
```diff
@@ -29,10 +30,11 @@ module.exports = function sendTraces(parentCtl) {
     return;
   }
 
-  cluster.on('fork', function(worker) {
+  master.on('fork', function(worker) {
     worker.on('message', function(msg) {
       if (msg.cmd !== 'agent:trace') return;
       debug('master received: %j', debug.json(msg.trace));
+      msg.pst = worker.startTime;
```
Comment (sam-github) [inline lib/traces.js]: Start time is defined by the parent, I take it? Not by the worker?

**Comment by sam-github** [comment]
LGTM, I  think, if you can confirm: the intention is for every message that has `pid/wid`, it will have `pid/wid/pst`? I think you may have missed `lib/metrics`.

**Comment by rmg** [comment]
Yes, the intention is for every `pid` to be paired with a `pst`.

As for metrics, they don't include the `pid`, it's just `{cmd: 'metrics', metrics: metrics}`.

File: lib/runctl.js [hunk 4]
```diff
@@ -150,9 +154,15 @@ function onRequest(req, callback) {
     cluster.disconnect();
 
   } else if (cmd === 'fork') {
-    var child = cluster.fork();
-    rsp.workerID = child.workerID;
-    rsp.processID = child.process.pid;
+    rsp = null;
+    cluster.fork().once('fork', function(worker) {
```
Comment (rmg) [inline lib/runctl.js]: @sam-github this is where the worker emitting a `fork` event is useful. (re: comment on strongloop/strong-cluster-control#49)

File: lib/traces.js (hunk 3 above)
Comment (rmg) [inline lib/traces.js]: Correct, it is based on when the master gets the `fork` event for a given worker.

File: lib/runctl.js (hunk 4 above)
Comment (sam-github) [inline lib/runctl.js]: I thought this was a pure-refactor, and you just didn't like assigning to rsp, now I understand better...

Please add a comment: the fork event is emitted *after* strong-cluster-control has extended it with some custom properties, which is why we can't simply use the fork() return value.

File: lib/runctl.js (hunk 1 above)
Comment (sam-github) [inline lib/runctl.js]: I understand why its here: consistency.

File: lib/status-wd.js (hunk 2 above)
Comment (sam-github) [inline lib/status-wd.js]: OK

File: lib/traces.js (hunk 3 above)
Comment (sam-github) [inline lib/traces.js]: ok

**Comment by sam-github** [comment]
metrics is 
```{cmd: 'metrics', metrics: {
  processes: {
     [wid]: { /* metrics */ } // pid,wid,pst can be assigned into this obj based on the wid
  }
}}
```

Othat than that, LGTM

**Pull request merged**